pipelines/google_sheets/google_sheets.py

```diff
@@ -0,0 +1,133 @@
+# Contains the main pipeline functions
+
+from typing import Any, Iterator, cast
+import dlt
+from dlt.common.configuration.specs import GcpClientCredentialsWithDefault
+from dlt.common.typing import DictStrAny
+from dlt.common.exceptions import MissingDependencyException
+from dlt.common.typing import TDataItems
+import logging
+import time
+
+try:
+    from data_validator import serial_date_to_datetime, process_url
+except ModuleNotFoundError:
+    from google_sheets.data_validator import process_url, serial_date_to_datetime
+
+try:
+    from apiclient.discovery import build, Resource
+except ImportError:
+    raise MissingDependencyException("Google Sheets Source", ["google1", "google2"])
+
+# constants
+SPREADSHEET_ID = "1NVxFQYRqrGmur_MeIc4m4ewToF802uy2ObC61HOCstU"
+SPREADSHEET_URL = "https://docs.google.com/spreadsheets/d/1NVxFQYRqrGmur_MeIc4m4ewToF802uy2ObC61HOCstU/edit#gid=0"
+
+
+# gets sheet values from a single spreadsheet preserving original typing for schema generation
+# TODO: write schema discovery to better handle dates
+# TODO: consider using https://github.com/burnash/gspread for spreadsheet discovery
+# TODO: add exceptions to the code in case HTTP requests fail
+
+# this function receives 2 arrays of tabular data inside a sheet. This will be processed into a schema that is later stored into a database table
+# it will then
+def process_range(sheet_val: Iterator[DictStrAny], sheet_val_sn: Iterator[DictStrAny]) -> Iterator[DictStrAny]:
+
+    # get headers - same for both arrays
+    headers = sheet_val[0]  # list
+    for row, row_sn in zip(sheet_val[1:], sheet_val_sn[1:]):
+        table_dict = {}
+        # process both rows and check for differences to spot dates
+        for val, val_sn, header in zip(row, row_sn, headers):
+            # check that they have different types and then process the dates
+            if type(val) != type(val_sn) and (isinstance(val_sn, int) or isinstance(val_sn, float)):
```

Comment:
make it more efficient by
1. not comparing types but makeing sure val is instance of string.
2. `isinstance` takes (int, float) to check multiple types, see docs